File: traversal.c
#include <stdio.h>
#include "traversal.h"

void inorder(struct Node *root)
{
    if (root != NULL)
    {
        inorder(root->left);
        printf("%c ", root->data);
        inorder(root->right);
    }
}

void preorder(struct Node *root)
{
    if (root != NULL)
    {
        printf("%c ", root->data);
        preorder(root->left);
        preorder(root->right);
    }
}

void postorder(struct Node *root)
{
    if (root != NULL)
    {
        postorder(root->left);
        postorder(root->right);
        printf("%c ", root->data);
    }
}

File: Traversal.java
public class Traversal {

    // inorder traversal
    public static String inorder(ExpressionTree.Node root) {
        if (root == null) return "";
        return inorder(root.left) + root.data + " " + inorder(root.right);
    }

    // preorder traversal
    public static String preorder(ExpressionTree.Node root) {
        if (root == null) return "";
        return root.data + " " + preorder(root.left) + preorder(root.right);
    }

    // postorder traversal
    public static String postorder(ExpressionTree.Node root) {
        if (root == null) return "";
        return postorder(root.left) + postorder(root.right) + root.data + " ";
    }
}
